DEVELOPMENT.md

```diff
@@ -115,3 +132,58 @@ bazel build third_party/chromium
 
 # Lastly, put the address returnd by the web server into the Chromium.
 ```
+
+### Debugging Angular UI Tests Locally
+
+tl;dr. Use `ibazel test` for regular work (supports limiting the test amount with
+`fit/fdescribe` and `console.log` usage) and `ibazel run` for karma console breakpoint
+debugging.
+
+1.  Just run all webapp tests. The job stops after finished. `console.log` is not
+    supported. Not handy on debugging.
+
+    ```sh
+    (tf)$ bazel test //tensorboard/webapp/...
+    ```
+
+    It runs the following tests:
+
+    ```sh
+    //tensorboard/webapp:karma_test_chromium-local PASSED in 42.3s
+    //tensorboard/webapp/feature_flag:karma_test_chromium-local PASSED in 5.8s
+    //tensorboard/webapp/hparams:karma_test_chromium-local PASSED in 5.7s
+    //tensorboard/webapp/notification_center:notification_center_test_chromium-local PASSED in 10.1s
+    //tensorboard/webapp/persistent_settings:karma_test_chromium-local PASSED in 10.9s
+    //tensorboard/webapp/settings:karma_test_chromium-local PASSED in 6.9s
+    ```
+
+2.  Using `ibazel` to auto detect the file changes and use target
+    `karma_test_chromium-local` for running on *webapp* tests. Practically we can add
+    `console.log` and `fit/fdescribe` (which areused to narrow down the test amount) and no need to restart the tests.
+
+    ```sh
+    (tf)$ ibazel test //tensorboard/webapp/... --test_output=all
+    ```
+
+    *   `--test_output=all`: for displaying number of tests if using '`fit`'.
+
+3.  To run on a specific test, we can change the target (with `chromium-local`
+    suffix). For example,
+
+    ```sh
+    //  Run webapp tests on `karma_test` target
+    (tf)$ ibazel test //tensorboard/webapp:karma_test_chromium-local
+
+    //  Run notification center tests on `notification_center_test` target
+    (tf)$ ibazel test //tensorboard/webapp/notification_center:notification_center_test_chromium-local
+    ```
+
+4.  For having a karma console to set break points for debugging purpose, use
+    `ibazel run`. Access the karma console at port `9876` (For example, `http://<YOUR_SERVER_ADDRESS>:9876/`) and click 'DEBUG' button, it pops up another page, where you have to look at the browser developer console to see the logs for debugging.
+
+    ```sh
+    (tf)$ ibazel run //tensorboard/webapp:karma_test_chromium-local
+    ```
+
+    However, with `ibazel run` the file watcher is glitchy on running the tests
+    when detecting changes. It shows '`a broken pipe`' in terminal. We need to terminate and restart the program manually.
```

Comment:
Given this limitation you may as well just recommend using `bazel run` instead of `ibazel` but up to you.